### PR TITLE
add sortable table header

### PR DESCRIPTION
**Pull request by simleap**
Pivotal: https://www.pivotaltracker.com/story/show/157724643

screenshot 2018-05-30 15 06 53

### CODE REVIEW

File: app/views/shared/_resources_table.html.haml [hunk 1]
```diff
@@ -2,9 +2,21 @@
...
       %th #
+      - exclude_columns = [:province, :actions, :level]
       - columns.each do |column|
-        %th
-          = resource_class.human_attribute_name(column.to_sym)
+        %th.th-sortable
+          - if column.in? exclude_columns
+            = resource_class.human_attribute_name(column.to_sym)
+          - else
+            - is_current_comlumn = (column.to_s == sort)
```
Comment (dwilkie) [inline app/views/shared/_resources_table.html.haml]: Typo. Should be `is_current_column`

File: app/controllers/dashboard/base_controller.rb [hunk 2]
```diff
@@ -13,7 +13,7 @@ class Dashboard::BaseController < ApplicationController
   before_action :authenticate_user!
   before_action :find_resource, only: %i[show edit update destroy]
 
-  helper_method :resource, :resources, :show_location, :current_account
+  helper_method :resource, :resources, :show_location, :current_account, :sort, :direction
```
Comment (samnang) [inline app/controllers/dashboard/base_controller.rb]: Let's call it `helper_method :sort_column, :sort_direction`

File: app/controllers/dashboard/base_controller.rb [hunk 3]
```diff
@@ -53,7 +53,7 @@ def destroy
 
   private
 
-  attr_reader :resource, :resources
+  attr_reader :resource, :resources, :sort, :direction
```
Comment (samnang) [inline app/controllers/dashboard/base_controller.rb]: I don't think we need attr_reader here.

```ruby
 def sort_column
    params[:sort] || "created_at"
  end
  
  def sort_direction
    %w[asc desc].include?(params[:direction]) ? params[:direction] : "desc"
  end
```

File: app/views/shared/_resources_table.html.haml [hunk 4]
```diff
@@ -2,9 +2,21 @@
   %thead.thead-light
     %tr
       %th #
+      - exclude_columns = [:province, :actions, :level]
```
Comment (samnang) [inline app/views/shared/_resources_table.html.haml]: `exclude_columns` option should be passed from the caller of this partial whenever they need it. This partial is used my many places, so we can't just hardcodes fields from a specific model.

`exclude_columns` option should be optional.

File: app/views/shared/_resources_table.html.haml [hunk 5]
```diff
@@ -2,9 +2,21 @@
...
+          - else
+            - is_current_column = (column.to_s == sort)
+            - is_current_asc = (direction == "asc")
+            - new_direction = (is_current_column && is_current_asc) ? "desc" : "asc"
+
+            = link_to url_for(action: "index", sort: column, direction: new_direction) do
+              = resource_class.human_attribute_name(column.to_sym)
+            %span.sortable-icons
+              = fa_icon "sort-up" if !is_current_column || is_current_asc
+              = fa_icon "sort-down" if !is_current_column || !is_current_asc
```
Comment (samnang) [inline app/views/shared/_resources_table.html.haml]: How about we refactor something like this:

```rb
  direction = (column == sort_column && sort_direction == "desc") ? "asc" : "desc"
  link_to url_for(sort: column, direction: direction) do
    = resource_class.human_attribute_name(column.to_sym)
  end

  %span.sortable-icons
    - if direction == "desc"
      = fa_icon "sort-up"
    - else
      = fa_icon "sort-asc"

```

File: spec/system/dashboard/users_spec.rb [hunk 6]
```diff
@@ -25,13 +24,22 @@
...
+        href: dashboard_users_path(sort: "invitation_accepted_at", direction: "asc")
+      )
+      expect(page).to have_link(
+        User.human_attribute_name(:last_sign_in_at),
+        href: dashboard_users_path(sort: "last_sign_in_at", direction: "asc")
+      )
+      expect(page).to have_link(
+        User.human_attribute_name(:created_at),
+        href: dashboard_users_path(sort: "created_at", direction: "asc")
+      )
```
Comment (samnang) [inline spec/system/dashboard/users_spec.rb]: To me, I think it's a bit too much and has little values on repeat all assertions here for each page and column like this.

I would rather test have just one scenario to verify if I could be able to click see the default sort, then I I click sort in the opposite direction then I see the result that I expect.

Another test should only verify which column that could be sortable by assertion with class selector something like this.

```
expect(page).to have_sortable_column(:email)
expect(page).to have_sortable_column(:last_sign_in_at)

def have_sortable_column(column)
  have_link(".sortable.#{column}")
end
```

File: app/controllers/dashboard/base_controller.rb (hunk 3 above)
Comment (samnang) [reply app/controllers/dashboard/base_controller.rb]: @simleap how about this?

File: db/schema.rb [hunk 7]
```diff
@@ -167,6 +188,24 @@
...
+
+  create_table "sensors", force: :cascade do |t|
+    t.jsonb "metadata", default: {}, null: false
+    t.bigint "account_id"
+    t.string "external_id", null: false
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+    t.index ["account_id", "external_id"], name: "index_sensors_on_account_id_and_external_id", unique: true
+    t.index ["account_id"], name: "index_sensors_on_account_id"
+  end
```
Comment (samnang) [inline db/schema.rb]: Why these changes in this PR?

File: app/views/shared/_resources_table.html.haml [hunk 8]
```diff
@@ -3,8 +3,23 @@
...
-        %th
-          = resource_class.human_attribute_name(column.to_sym)
+        %th.th-sortable
+          - if (defined? unsortable_columns) && (column.in? unsortable_columns)
+            .text
+              = resource_class.human_attribute_name(column.to_sym)
+          - else
+            - is_current_column = (column.to_s == sort_column)
+            - is_current_asc = (sort_direction == "asc")
+            - new_direction = (is_current_column && is_current_asc) ? "desc" : "asc"
```
Comment (samnang) [inline app/views/shared/_resources_table.html.haml]: We want the default to be `desc`, right? In this case, for other columns will be in `asc`.

Comment (simleap) [reply app/views/shared/_resources_table.html.haml]: In this case, we want default sort by created_at with newest right? and this shows newest first.

File: app/controllers/dashboard/base_controller.rb (hunk 3 above)
Comment (samnang) [reply app/controllers/dashboard/base_controller.rb]: @simleap we don't need these attr_reader `:sort, :direction`, right?

File: app/views/shared/_resources_table.html.haml [hunk 9]
```diff
@@ -3,8 +3,23 @@
     %tr
       %th #
       - columns.each do |column|
-        %th
-          = resource_class.human_attribute_name(column.to_sym)
+        %th.th-sortable
+          - if (defined? unsortable_columns) && (column.in? unsortable_columns)
```
Comment (samnang) [inline app/views/shared/_resources_table.html.haml]: Here don't no use `defined?` you can use `Array(local_assigns[:unsortable_columns])` instead.

e.g. 

```ruby
-if column.in?(Array(local_assigns[:unsortable_columns]))
```
